test/object_store/conftest.py

```diff
@@ -45,28 +45,26 @@ def ssl(request):
     yield request.config.getoption('--ssl')
 
 
-@contextmanager
-def _maybe_remove_tempdir_on_complete(request, tempdir):
-    if pytest.version_tuple > (7, 3, 0):
-        yield
-        return
-    # tmp_path_retention_count was introduced in pytest 7.3.0, so we have
-    # to swing this by ourselves when using lower version of pytest. see
-    # https://docs.pytest.org/en/7.3.x/changelog.html#pytest-7-3-0-2023-04-08
-    tests_failed = request.session.testsfailed
-    yield
-    if tests_failed == request.session.testsfailed:
-        # not tests failed after this seesion, so remove the tempdir
-        shutil.rmtree(tempdir)
+def _remove_all_but(tempdir, to_preserve):
+    orig_fn = os.path.join(tempdir, to_preserve)
+    try:
+        with tempfile.TemporaryDirectory() as backup_tempdir:
+            backup_fn = os.path.join(backup_tempdir, to_preserve)
+            shutil.move(orig_fn, backup_fn)
+            shutil.rmtree(tempdir)
+            os.mkdir(tempdir)
+            shutil.move(backup_fn, orig_fn)
+    except FileNotFoundError:
```

Comment:
This exception handling makes me worry. Apparently it's aimed at `to_preserve` missing and `shutil.move()` throws. In that case the whole `tempdir` is left not removed, which is, likely, not what should happen.